### PR TITLE
Show a path to make ValuesSourceRegistry immutable

### PR DESCRIPTION
**Pull request by nik9000**
Most of the stuff we build at node startup is immutable but
ValuesSourceRegistry isn't. This implements some of the pieces that we'd
need to make it immutable. It also begins to remove it from
`SearchPlugin` in an effort to keep the plugin interface as small as
possible.

### CODE REVIEW

**Comment by nik9000** [comment]
@not-napoleon, have a look at what I came up with for this. I think it is an improvement, but I don't claim that it is *right*.

**Review by not-napoleon — COMMENTED** [review]
I'm 100% with you that `ValuesSourceRegistry` should be immutable.  I think this is a reasonable path to that.  I'm a little frustrated that it means we'll have to redo every wire up that we've already done, but I guess there's no help for that.  I do feel like we need a clear and easy-to-follow pattern for how to register an aggregator, which should be the same if it's a core aggregator or a plugin aggregator.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java [hunk 1]
```diff
@@ -59,15 +71,10 @@
      */
     public synchronized void register(String aggregationName, Predicate<ValuesSourceType> appliesTo,
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java]: Sure!

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java [hunk 2]
```diff
@@ -169,8 +176,8 @@ public ValuesSourceType getValuesSourceType(MappedFieldType fieldType, String ag
         }
     }
 
-    private static <K, V> Map copyAndAdd(Map<K, V>  source, Map.Entry<K, V>  newValue) {
-        Map.Entry[] entries;
+    private static <K, V> Map<K, V> copyAndAdd(Map<K, V>  source, Map.Entry<K, V>  newValue) {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java]: I believe we won't.

[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java]: I could revert my changes to it. I was sort of trying to make my IDE stop highlighting things.

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java [hunk 3]
```diff
@@ -176,12 +178,16 @@ private static void registerFieldTypes(SearchContext searchContext, MapperServic
         }
     }
 
-    @BeforeClass
-    public static void initValuesSourceRegistry() {
-        SearchModule searchModule = new SearchModule(Settings.EMPTY, List.of());
+    @Before
```
[the remark this answers is not in the document]
Comment (nik9000) [reply test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java]: We need `plugins` to be override-able so we can plug extensions into the test.

File: server/src/main/java/org/elasticsearch/plugins/SearchPlugin.java [hunk 4]
```diff
@@ -260,6 +264,7 @@ public QuerySpec(String name, Writeable.Reader<T> reader, QueryParser<T> parser)
      */
     class AggregationSpec extends SearchExtensionSpec<AggregationBuilder, ContextParser<String, ? extends AggregationBuilder>> {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/plugins/SearchPlugin.java]: Maybe! Or maybe I should spend a little more time stopping it from leaking....

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/AvgAggregationBuilder.java [hunk 5]
```diff
@@ -24,30 +24,33 @@
...
-    public static final ObjectParser<AvgAggregationBuilder, String> PARSER = ObjectParser.fromBuilder(NAME, AvgAggregationBuilder::new);
+    private static final ObjectParser<AvgAggregationBuilder, String> PARSER = ObjectParser.fromBuilder(NAME, AvgAggregationBuilder::new);
     static {
         ValuesSourceAggregationBuilder.declareFields(PARSER, true, true, false);
     }
 
-    public static void registerAggregators(ValuesSourceRegistry valuesSourceRegistry) {
-        AvgAggregatorFactory.registerAggregators(valuesSourceRegistry);
-    }
+    public static final AggregationSpec SPEC = new AggregationSpec(NAME, AvgAggregationBuilder::new, PARSER)
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/AvgAggregationBuilder.java]: Yeah, they need the license check and the accounting. Both of those things are very "plugin" and not very "aggregation" so I figured a method to build the spec made sense there. But I prefer a static spec. Though, to be honest, maybe we don't need a static spec and a method is always better. It isn't like we need to keep the spec around.

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (not-napoleon) [reply test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:181]: oh, I see.  So it's not that it needs. to run per-test, but that it needs to not be static, which means in practice it needs to be an `@Before` not an `@BeforeClass` annotation.  I'd say we should put a comment to note that, but I don't even know where we'd put it to be helpful.  :-/

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/AvgAggregationBuilder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/AvgAggregationBuilder.java:51]: I'd feel better about this if there was one pattern we used in both places.

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:181]: We do it all over the place, I think.

**Comment by nik9000** [comment]
@not-napoleon, check it out now!

**Comment by nik9000** [comment]
I talked to @not-napoleon and we think we'll go the direction with this and pass a builder into the plugins - that way changing them to accept the builder will be quite simple.

**Pull request closed without merging**